### PR TITLE
net: dns: add dns_resolve_reconfigure() API

### PR DESCRIPTION
**Pull request by mniestroj**
So far there was no dedicated mechanism for replacing DNS servers with
new list. Add dns_resolve_reconfigure() API that allows to achieve that
in a thread-safe manner.

Update all components that reconfigured DNS server list (like DHCPv4,
PPP, ESP-AT WiFi and hl7800 modem) to use that API.

### CODE REVIEW

File: subsys/net/lib/dns/resolve.c [hunk 1]
```diff
@@ -1271,10 +1302,14 @@ static int dns_resolve_close_locked(struct dns_resolve_context *ctx)
...
 	int i;
 
-	if (!ctx->is_used) {
+	if (ctx->state != DNS_RESOLVE_CONTEXT_ACTIVE) {
 		return -ENOENT;
 	}
 
+	ctx->state = DNS_RESOLVE_CONTEXT_DEACTIVATING;
+
+	k_mutex_unlock(&ctx->lock);
```
Comment (jukkar) [inline subsys/net/lib/dns/resolve.c:1311]: Why do we unlock, after all for loop is still accessing the ctx fields.

File: subsys/net/ip/dhcpv4.c [hunk 2]
```diff
@@ -764,17 +764,8 @@ static bool dhcpv4_parse_options(struct net_pkt *pkt,
 			}
```
Comment (jukkar) [inline subsys/net/ip/dhcpv4.c:764]: typo in commit `net: dhcpv4: use dns_resolve_configure() API` subject, s/configure/reconfigure/
same comment to all the other commits in this PR that have similar text

**Review by jukkar — COMMENTED** [review]
Looks good, minor comments

File: subsys/net/lib/dns/resolve.c [hunk 3]
```diff
@@ -331,14 +329,29 @@ int dns_resolve_init(struct dns_resolve_context *ctx, const char *servers[],
...
+	if (!ctx) {
+		return -ENOENT;
+	}
+
+	(void)memset(ctx, 0, sizeof(*ctx));
+
+	(void)k_mutex_init(&ctx->lock);
+	ctx->state = DNS_RESOLVE_CONTEXT_INACTIVE;
+
+	return dns_resolve_init_locked(ctx, servers, servers_sa);
```
[the remark this answers is not in the document]
Comment (mniestroj) [reply subsys/net/lib/dns/resolve.c:352]: I was assuming that `dns_resolve_init()` is called only once during system init. This makes sure that no other thread will use DNS contexts yet, so locking is not needed. Should we make such comment in code?

File: subsys/net/lib/dns/resolve.c (hunk 1 above)
Comment (mniestroj) [reply subsys/net/lib/dns/resolve.c:1334]: Good question. Added following comment:
```
	/*
	 * ctx->net_ctx is never used in "deactivating" state. Additionally
	 * following code is guaranteed to be executed only by one thread at a
	 * time, due to required "active" -> "deactivating" state change. This
	 * means that it is safe to put net_ctx with mutex released.
	 *
	 * Released mutex will prevent lower networking layers from deadlock
	 * when calling cb_recv() (which acquires ctx->lock) just before closing
	 * network context.
	 */
	k_mutex_unlock(&ctx->lock);
```

File: subsys/net/ip/dhcpv4.c (hunk 2 above)
Comment (mniestroj) [reply subsys/net/ip/dhcpv4.c:764]: Fixed, thanks!

File: subsys/net/lib/dns/resolve.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (jukkar) [reply subsys/net/lib/dns/resolve.c:352]: Sure, that was my though too (that we call this only once in the system). It is just that we are calling a function that is suppose to be called with things locked. Perhaps it make sense to add a comment here, then if someone reads the code it is clear what is the intention here.

File: subsys/net/lib/dns/resolve.c (hunk 1 above)
Comment (jukkar) [reply subsys/net/lib/dns/resolve.c:1334]: ok

File: subsys/net/lib/dns/resolve.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (mniestroj) [reply subsys/net/lib/dns/resolve.c:355]: Done.

**Comment by mniestroj** [comment]
Compliance check fails with:
```
WARNING:STATIC_CONST_CHAR_ARRAY: char * array declaration might be better as static const
#471: FILE: tests/net/lib/dns_addremove/src/main.c:445:
+	const char *dns_servers_str[] = { DNS_NAME_IPV4, NULL };
```
after adding `static` there is still a warning:
```
WARNING:STATIC_CONST_CHAR_ARRAY: static const char * array should probably be static const char * const
#34: FILE: tests/net/lib/dns_addremove/src/main.c:445:
+	static const char *dns_servers_str[] = { DNS_NAME_IPV4, NULL };
```
and after following that advice code no longer compiles:
```
/home/macius/grinn/seas-nve/project/zephyr/tests/net/lib/dns_addremove/src/main.c: In function 'test_dns_reconfigure_callback':
/home/macius/grinn/seas-nve/project/zephyr/tests/net/lib/dns_addremove/src/main.c:449:33: error: passing argument 2 of 'dns_resolve_init' discards 'const' qualifier from pointer target type [-Werror=discarded-qualifiers]
  449 |  ret = dns_resolve_init(dnsCtx, dns_servers_str, NULL);
      |                                 ^~~~~~~~~~~~~~~
In file included from /home/macius/grinn/seas-nve/project/zephyr/tests/net/lib/dns_addremove/src/main.c:25:
/home/macius/grinn/seas-nve/project/zephyr/include/net/dns_resolve.h:277:20: note: expected 'const char **' but argument is of type 'const char * const*'
  277 |        const char *dns_servers_str[],
      |        ~~~~~~~~~~~~^~~~~~~~~~~~~~~~~
```
so I left `const char *dns_servers_str[]` to be in sync with other test in the same file.

**Comment by jukkar** [comment]
@mniestroj The build checks are in weird state as they report errors even if offending code is not found any longer in the code. Could you rebase against latest upstream and re-submit, perhaps that would remove some of the failures.

**Comment by jukkar** [comment]
The compliance check seems to be false positive.
The `clang-build(4)` has this issue
```
tests/net/lib/dns_packet/src/main.c:1066:7: error: no member named 'is_used' in 'struct dns_resolve_context'
        ctx->is_used = true;
        ~~~  ^
1 error generated.
```

**Review by jukkar — APPROVED** [review]

**Review by rerickson1 — APPROVED** [review]

**Comment by jukkar** [comment]
@galak / @nashif there is a false positive in checkpatch, could you override it?

**Review by rlubos — APPROVED** [review]

**Pull request merged**